File: AwsCryptographicMaterialProviders/runtimes/net/Extern/StormTrackingCMC.cs
// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

using software.amazon.cryptography.primitives.internaldafny.types;

namespace software.amazon.cryptography.internaldafny.StormTrackingCMC
{
  public partial class StormTrackingCMC : software.amazon.cryptography.materialproviders.internaldafny.types.ICryptographicMaterialsCache
  {
    public StormTracker_Compile.StormTracker wrapped;

    public StormTrackingCMC(StormTracker_Compile.StormTracker cmc)
    {
      this.wrapped = cmc;
    }

    [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.Synchronized)]
    public Wrappers_Compile._IResult<_System._ITuple0, software.amazon.cryptography.materialproviders.internaldafny.types._IError> PutCacheEntry(software.amazon.cryptography.materialproviders.internaldafny.types._IPutCacheEntryInput input)
    {
      return PutCacheEntry_k(input);
    }

    [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.Synchronized)]
    public Wrappers_Compile._IResult<_System._ITuple0, software.amazon.cryptography.materialproviders.internaldafny.types._IError> UpdateUsageMetadata(software.amazon.cryptography.materialproviders.internaldafny.types._IUpdateUsageMetadataInput input)
    {
      return UpdateUsageMetadata_k(input);
    }

    // NOT synchronized, as some sleeping might be involved
    public Wrappers_Compile._IResult<software.amazon.cryptography.materialproviders.internaldafny.types._IGetCacheEntryOutput, software.amazon.cryptography.materialproviders.internaldafny.types._IError> GetCacheEntry(software.amazon.cryptography.materialproviders.internaldafny.types._IGetCacheEntryInput input)
    {
      return GetCacheEntry_k(input);
    }

    [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.Synchronized)]
    public Wrappers_Compile._IResult<_System._ITuple0, software.amazon.cryptography.materialproviders.internaldafny.types._IError> DeleteCacheEntry(software.amazon.cryptography.materialproviders.internaldafny.types._IDeleteCacheEntryInput input)
    {
      return DeleteCacheEntry_k(input);
    }

    // This is the synchronization for GetCacheEntry and GetCacheEntry_k
    [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.Synchronized)]
    public Wrappers_Compile._IResult<StormTracker_Compile._ICacheState, software.amazon.cryptography.materialproviders.internaldafny.types._IError>
      GetFromCacheInner(software.amazon.cryptography.materialproviders.internaldafny.types._IGetCacheEntryInput input) {
      return wrapped.GetFromCache(input);
    }

    // NOT synchronized, as some sleeping might be involved
    public Wrappers_Compile._IResult<software.amazon.cryptography.materialproviders.internaldafny.types._IGetCacheEntryOutput, software.amazon.cryptography.materialproviders.internaldafny.types._IError> GetCacheEntry_k(software.amazon.cryptography.materialproviders.internaldafny.types._IGetCacheEntryInput input)
    {
      while (true) {
        Wrappers_Compile._IResult<StormTracker_Compile._ICacheState, software.amazon.cryptography.materialproviders.internaldafny.types._IError>
          result = GetFromCacheInner(input);
        if (result.is_Failure) {
          return Wrappers_Compile.Result<software.amazon.cryptography.materialproviders.internaldafny.types._IGetCacheEntryOutput, software.amazon.cryptography.materialproviders.internaldafny.types._IError>
            .create_Failure((result).dtor_error);
        } else if (result.dtor_value.is_Full) {
          return Wrappers_Compile.Result<software.amazon.cryptography.materialproviders.internaldafny.types._IGetCacheEntryOutput, software.amazon.cryptography.materialproviders.internaldafny.types._IError>
            .create_Success(result.dtor_value.dtor_data);
        } else if (result.dtor_value.is_EmptyFetch) {
          return Wrappers_Compile.Result<software.amazon.cryptography.materialproviders.internaldafny.types._IGetCacheEntryOutput, software.amazon.cryptography.materialproviders.internaldafny.types._IError>
              .create_Failure(software.amazon.cryptography.materialproviders.internaldafny.types.Error
                  .create_EntryDoesNotExist(Dafny.Sequence<char>.FromString("Entry does not exist")));
        } else {
          Thread.Sleep(50);
        }
      }
    }
    
    [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.Synchronized)]
    public Wrappers_Compile._IResult<_System._ITuple0, software.amazon.cryptography.materialproviders.internaldafny.types._IError> PutCacheEntry_k(software.amazon.cryptography.materialproviders.internaldafny.types._IPutCacheEntryInput input)
    {
      return this.wrapped.PutCacheEntry(input);
    }

    [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.Synchronized)]
    public Wrappers_Compile._IResult<_System._ITuple0, software.amazon.cryptography.materialproviders.internaldafny.types._IError> DeleteCacheEntry_k(software.amazon.cryptography.materialproviders.internaldafny.types._IDeleteCacheEntryInput input)
    {
      return this.wrapped.DeleteCacheEntry(input);
    }

    [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.Synchronized)]
    public Wrappers_Compile._IResult<_System._ITuple0, software.amazon.cryptography.materialproviders.internaldafny.types._IError> UpdateUsageMetadata_k(software.amazon.cryptography.materialproviders.internaldafny.types._IUpdateUsageMetadataInput input)
    {
      return this.wrapped.UpdateUsageMetadata(input);
    }
  }
}

File: AwsCryptographicMaterialProviders/runtimes/java/src/main/java/software/amazon/cryptography/internaldafny/StormTrackingCMC/StormTrackingCMC.java
package software.amazon.cryptography.internaldafny.StormTrackingCMC;

import static software.amazon.cryptography.internaldafny.StormTrackingCMC._ExternBase___default.CreateGetCacheEntryFailure;
import static software.amazon.cryptography.internaldafny.StormTrackingCMC._ExternBase___default.CreateGetCacheEntrySuccess;

import StormTracker_Compile.CacheState;
import StormTracker_Compile.StormTracker;
import software.amazon.cryptography.materialproviders.internaldafny.*;
import software.amazon.cryptography.materialproviders.internaldafny.types.*;

@SuppressWarnings({ "unchecked", "deprecation" })
public class StormTrackingCMC
  implements
    software.amazon.cryptography.materialproviders.internaldafny.types.ICryptographicMaterialsCache {

  private StormTracker wrapped;

  public StormTrackingCMC(StormTracker wrapped) {
    (this).wrapped = wrapped;
  }

  public synchronized Wrappers_Compile.Result<
    dafny.Tuple0,
    software.amazon.cryptography.materialproviders.internaldafny.types.Error
  > PutCacheEntry(
    software.amazon.cryptography.materialproviders.internaldafny.types.PutCacheEntryInput input
  ) {
    return wrapped.PutCacheEntry(input);
  }

  public synchronized Wrappers_Compile.Result<
    dafny.Tuple0,
    software.amazon.cryptography.materialproviders.internaldafny.types.Error
  > UpdateUsageMetadata(
    software.amazon.cryptography.materialproviders.internaldafny.types.UpdateUsageMetadataInput input
  ) {
    return wrapped.UpdateUsageMetadata(input);
  }

  // NOT synchronized, as some sleeping might be involved
  public Wrappers_Compile.Result<
    software.amazon.cryptography.materialproviders.internaldafny.types.GetCacheEntryOutput,
    software.amazon.cryptography.materialproviders.internaldafny.types.Error
  > GetCacheEntry(
    software.amazon.cryptography.materialproviders.internaldafny.types.GetCacheEntryInput input
  ) {
    return GetCacheEntry_k(input);
  }

  public synchronized Wrappers_Compile.Result<
    dafny.Tuple0,
    software.amazon.cryptography.materialproviders.internaldafny.types.Error
  > DeleteCacheEntry(
    software.amazon.cryptography.materialproviders.internaldafny.types.DeleteCacheEntryInput input
  ) {
    return wrapped.DeleteCacheEntry(input);
  }

  public synchronized Wrappers_Compile.Result<
    dafny.Tuple0,
    software.amazon.cryptography.materialproviders.internaldafny.types.Error
  > PutCacheEntry_k(
    software.amazon.cryptography.materialproviders.internaldafny.types.PutCacheEntryInput input
  ) {
    return wrapped.PutCacheEntry(input);
  }

  public synchronized Wrappers_Compile.Result<
    dafny.Tuple0,
    software.amazon.cryptography.materialproviders.internaldafny.types.Error
  > UpdateUsageMetadata_k(
    software.amazon.cryptography.materialproviders.internaldafny.types.UpdateUsageMetadataInput input
  ) {
    return wrapped.UpdateUsageMetadata(input);
  }

  // This is the synchronization for GetCacheEntry and GetCacheEntry_k
  public synchronized Wrappers_Compile.Result<
    CacheState,
    software.amazon.cryptography.materialproviders.internaldafny.types.Error
  > GetFromCacheInner(
    software.amazon.cryptography.materialproviders.internaldafny.types.GetCacheEntryInput input
  ) {
    return wrapped.GetFromCache(input);
  }

  // NOT synchronized, because we sleep. Calls GetFromCache which IS synchronized.
  public Wrappers_Compile.Result<
    software.amazon.cryptography.materialproviders.internaldafny.types.GetCacheEntryOutput,
    software.amazon.cryptography.materialproviders.internaldafny.types.Error
  > GetCacheEntry_k(
    software.amazon.cryptography.materialproviders.internaldafny.types.GetCacheEntryInput input
  ) {
    while (true) {
      Wrappers_Compile.Result<
        CacheState,
        software.amazon.cryptography.materialproviders.internaldafny.types.Error
      > result = GetFromCacheInner(input);
      if (result.is_Failure()) {
        return CreateGetCacheEntryFailure((result).dtor_error());
      } else if (result.dtor_value().is_Full()) {
        return CreateGetCacheEntrySuccess(result.dtor_value().dtor_data());
      } else if (result.dtor_value().is_EmptyFetch()) {
        return CreateGetCacheEntryFailure(
          software.amazon.cryptography.materialproviders.internaldafny.types.Error.create_EntryDoesNotExist(
            dafny.DafnySequence.asString("Entry does not exist")
          )
        );
      } else {
        try {
          Thread.sleep(wrapped.sleepMilli);
        } catch (Exception e) {}
      }
    }
  }

  public synchronized Wrappers_Compile.Result<
    dafny.Tuple0,
    software.amazon.cryptography.materialproviders.internaldafny.types.Error
  > DeleteCacheEntry_k(
    software.amazon.cryptography.materialproviders.internaldafny.types.DeleteCacheEntryInput input
  ) {
    return wrapped.DeleteCacheEntry(input);
  }

  @Override
  public java.lang.String toString() {
    return "StormTracker_Compile.StormTrackerCMC";
  }
}
